projects/plugins/boost/app/assets/src/js/features/critical-css/local-generator/local-generator-provider.tsx

```diff
@@ -0,0 +1,129 @@
+import { ReactNode, createContext, useContext, useEffect, useState } from 'react';
+import {
+	calculateCriticalCssProgress,
+	criticalCssErrorState,
+	useCriticalCssState,
+	useProxyNonce,
+	useSetProviderCssAction,
+	useSetProviderErrorsAction,
+	useRegenerateCriticalCssAction,
+} from '../lib/stores/critical-css-state';
+import { runLocalGenerator } from '../lib/generate-critical-css';
+import { CriticalCssErrorDetails } from '../lib/stores/critical-css-state-types';
+
+type LocalGeneratorContext = {
+	abortController?: AbortController;
+	setAbortController: ( controller: AbortController | undefined ) => void;
```

Comment:
I'm probably missing context and knowledge around React context and effects, but bear with me - 

This feels a bit difficult to follow AbortController around. Is there a way we can use the AbortController API fully, leveraging `controller.signal.aborted`?

Something like this: https://medium.com/@icjoseph/using-react-to-understand-abort-controllers-eb10654485df
This is using AbortController temporarily right within the `useEffect` - the controller never leaves the effect, that way the rest of the application doesn't have to know that an AbortController exists - all the application (or context in this case) knows is that CSS is either generating or it isn't.

If we did that, `LocalCriticalCssGeneratorProvider` could then have state -> `isRunning` or `isGenerating` that's always a boolean. `providerProgress` can then also be only a number, because it doesn't matter if the value is 0 or 100 if `isRunning` is false (dropping the need for undefined), although that's not a big deal - `undefined` is fine if you think it's better call it `undefined` when nothing is being generated.

Reply:
This is one of those things where I can see good arguments either way. I see your point about keeping the context simple, but I also felt that using the abortcontroller as a proxy for whether or not the process is running saved me from having to manage a separate boolean.

On the balance of things, I've gone with your suggestion in c190598ced.